### PR TITLE
Add read, read_string, and write functions to std::fs

### PR DESCRIPTION
**Pull request by SimonSapin**
New APIs in `std::fs`:

```rust
pub fn read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> { … }
pub fn read_string<P: AsRef<Path>>(path: P) -> io::Result<String> { … }
pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> io::Result<()> { ... }
```

(`read_string` is based on `read_to_string` and so returns an error on non-UTF-8 content.)

Before:

```rust
use std::fs::File;
use std::io::Read;

let mut bytes = Vec::new();
File::open(filename)?.read_to_end(&mut bytes)?;
do_something_with(bytes)
```

After:

```rust
use std::fs;

do_something_with(fs::read(filename)?)
```

### CODE REVIEW

**Comment by ollie27** [comment]
How about a version which reads into a `String` like #34857?

File: src/libstd/fs.rs
Comment (mbrubeck) [inline src/libstd/fs.rs]: This could look up the file size first and use `Vec::with_capacity`. It might mean as extra system call when reading a tiny file, but it is a [huge speedup][1] (up to 2× on my system) when reading large files.

[1]: https://github.com/TeXitoi/benchmarksgame-rs/issues/43

**Comment by SimonSapin** [comment]
I’m unsure about the naming. @mbrubeck suggests on IRC `read` and `write` free functions in the `std::fs` module.

File: src/libstd/fs.rs
Comment (SimonSapin) [reply src/libstd/fs.rs]: Maybe `read_to_end` should be responsible for that?

**Comment by jminer** [comment]
I've wanted these functions in Rust for a while. In .NET, there is [`File.ReadAllBytes`](https://msdn.microsoft.com/en-us/library/system.io.file.readallbytes(v=vs.110).aspx), [`File.WriteAllBytes`](https://msdn.microsoft.com/en-us/library/system.io.file.writeallbytes(v=vs.110).aspx), [`File.ReadAllText`](https://msdn.microsoft.com/en-us/library/ms143368(v=vs.110).aspx), [`File.WriteAllText`](https://msdn.microsoft.com/en-us/library/ms143375(v=vs.110).aspx), and [`File.AppendAllText`](https://msdn.microsoft.com/en-us/library/ms143356(v=vs.110).aspx) and . The .NET names have always seemed clear to me. I like that .NET uses "all" in the names to signal they read or write the whole file, compared to other read and write functions that don't.

**Comment by SimonSapin** [comment]
Alright, I pushed a couple new commits to switch to free functions and add reading to `String`. See updated PR message.

**Comment by euclio** [comment]
Why `read_utf8` over `read_string`?

**Comment by SimonSapin** [comment]
@euclio Because we’re reading bytes and decoding/validating them as UTF-8. Which happens to be the internal encoding of `String`, but that’s incidental as far as this function is concerned.

I my opinion it was a mistake for `Read::read_to_string` to have a name that does not include `utf8`. We do include it in `str::from_utf8`, `String::from_utf8`, `String::from_utf8_lossy`, `char::decode_utf8`.

File: src/libstd/fs.rs
Comment (mbrubeck) [reply src/libstd/fs.rs]: Specialization of `read_to_end` is a good idea. Filed #45851.

Note that callers of `read_to_end` can already avoid quadratic behavior by passing in a pre-allocated buffer.  That workaround isn't available with the functions in this PR.

**Comment by bluetech** [comment]
I'm not so sure this is a good idea. It combines the errors of two distinct operations. Also, if doing something like `fs::write(mypath)` + `fs::read(mypath)`, the path is opened two times, but in the meantime the file might have been deleted, etc. 

For `fs::write`, to me at least, it is not entirely clear from the name alone whether it truncates or overwrites (or even appends).

I assume the rationale is just making things easier? If so, the write case is already pretty short (and explicit):

```rust
File::create(path)?.write_all(contents)
```

For the read cases, maybe we can add instance methods instead? e.g.

```rust
File::open(path)?.slurp()
File::open(path)?.slurp_utf8()
```

This also seems short to me, not that far from Python's `open(path).read()`.

**Comment by sfackler** [comment]
> I'm not so sure this is a good idea. It combines the errors of two distinct operations.

The entire point of these functions is for the 90% use case where you don't care about distinguishing between those distinct operations.

**Comment by JordiPolo** [comment]
To me, given we have
```
File::open(filename)?.read_to_end(&mut bytes)?;
```
then
```
File::open(filename)?.read()?;
```
Seems easier to remember (do I want to reuse a buffer or just read and get something) than 
```
fs::read(filename)?
```
which is kind of stop thinking of files first, now thing of some other concept instead.

**Comment by scottmcm** [comment]
Like @jminer, I'm a big fan of these in .Net, and would love to have them in Rust.  They're rarely the absolute most optimal way to do whatever, but they're so convenient that I end up using them more than the "proper" way overall, since they're super-helpful in tests, adhoc tools, etc.

I like the name saying `utf8` explicitly, leaving space for a crate to make an encoding-adaptive version.

**Comment by SimonSapin** [comment]
@bluetech As @sfackler said often you don’t care about distinguishing these two errors, and when you do the separate functions/methods are still available.

I agree that `File::create(path)?.write_all(contents)?` is less bad than the read case that requires an intermediate variable, but it still requires importing the `std::io::Write` trait separately. These functions can be useful for example for quick prototyping, where even that can be additional friction.

As to `fs::write` followed immediately by `fs::read`, yes you wouldn’t do that in real code. You only see it in this diff because there’s a test for these functions themselves.

File: src/libstd/fs.rs
Comment (SimonSapin) [reply src/libstd/fs.rs]: I’ve opened a PR with a prototype: https://github.com/rust-lang/rust/pull/45928

**Comment by shepmaster** [comment]
Ping from triage @dtolnay — will you be able to spend some time on this?

**Comment by dtolnay** [comment]
I like this and I agree there is value in providing simple conveniences for the 90% use case where you don't care why reading from a file failed, whether you failed to open the file or failed to read bytes or the bytes weren't UTF-8. This resolves the idea Brian expressed in #34857 -- "be more aggressive about identifying and adding simple ergonomic improvements."

This is a significant enough new API that I would appreciate a few more eyes.

@rfcbot fcp merge

**Comment by sfackler** [comment]
@rfcbot concern write

I'm definitely on board with read and read_utf8, but I'm a worried about write due to the ambiguity around how it would behave with respect to file creation and truncation.

**Comment by BurntSushi** [comment]
I like these methods as well. I would be OK with `write` as-is, since I think the documented behavior is exactly what I'd expect. I do not feel strongly though!

**Comment by Ixrec** [comment]
Considering .NET has both File.WriteAllText and File.AppendAllText, we could consider also adding `std::fs::append` if we want to make it a little more obvious that `std::fs::write` does truncation.

**Comment by alexcrichton** [comment]
@rfcbot concern naming

I think I understand the rationale for `read_utf8` as a name but I feel like that just gets confusing if you're otherwise used to `Read`-the-trait. This is a case where I feel like consistency may buy us more than being extra explicit.

**Comment by SimonSapin** [comment]
I’ve already explained my preference for including `utf8` in the name, so I suppose the libs team should decide.

**Comment by aturon** [comment]
Very happy to see these coming! I do agree though that `read_utf8` would be better as `read_string`, for a few reasons:

- It's likelier to stand out if you're trying to get `String` output
- We already strongly emphasize that strings are utf8
- Consistency with `read`

**Comment by dtolnay** [comment]
I would lean toward `read_string` as well. In https://github.com/rust-lang/rust/pull/45837#issuecomment-342583259 you cited the following functions as the reason for the `read_utf8` name:

- [`str::from_utf8`](https://doc.rust-lang.org/std/str/fn.from_utf8.html), &[u8] ⟶ &str
- [`String::from_utf8`](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8), Vec\<u8> ⟶ String
- [`String::from_utf8_lossy`](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy), &[u8] ⟶ Cow\<str>
- [`char::decode_utf8`](https://doc.rust-lang.org/std/char/fn.decode_utf8.html), Iterator\<u8> ⟶ Iterator\<char>

I agree that there is symmetry in this name.

- [`fs::read_utf8`](#), bytes on the filesystem ⟶ String

But notice how in the first four names there is both "what you give" (bytes that are expected to be utf8) and "what you get" (str/String/char). In the name here, there is only "what you give" (bytes on the filesystem that are expected to be utf8). With `fs::read_string` there is both: "what you give" is a filesystem resource and "what you get" is a string (which are always utf8).

I believe expressing both "what you give" and "what you get" concisely is more valuable and contributes more to code that is easy to read and easy to understand, than being explicit about utf8 in this situation.

**Comment by dtolnay** [comment]
Regarding the ambiguity around how `fs::write` handles file creation and truncation, I agree with https://github.com/rust-lang/rust/pull/45837#issuecomment-345478537 that this is designed for the common case and does what you want in the common case, like the [.NET](https://msdn.microsoft.com/en-us/library/ms143375(v=vs.110).aspx) equivalent. I believe the documentation is sufficiently explicit.

File: src/libstd/fs.rs
Comment (mbrubeck) [inline src/libstd/fs.rs]: For convenience, should the second argument be `&V where V: AsRef<[u8]>` so that, for example, types like `String` and `&str` can be passed without `.as_bytes()`?

Comment (SimonSapin) [reply src/libstd/fs.rs]: Good idea. Done. (With `V` rather than `&V`, since `&T` implements `AsRef` if `T` does.)

**Comment by dtolnay** [comment]
@SimonSapin do the naming rationales in https://github.com/rust-lang/rust/pull/45837#issuecomment-346162867 and https://github.com/rust-lang/rust/pull/45837#issuecomment-347260204 affect your preference?

@sfackler do you think the ambiguity around fs::write can be sufficiently addressed in documentation, or do you think we need a different name / signature that is less ambiguous?

**Comment by alexcrichton** [comment]
@rfcbot resolved naming

**Comment by alexcrichton** [comment]
Er actually I thought these were renamed already, I think I'm still waiting on the response to https://github.com/rust-lang/rust/pull/45837#issuecomment-348814455

**Comment by scottmcm** [comment]
Is taking `AsRef<[u8]>` in `write` better than just having a a `write_string` that's symmetric with `read_string`?  Is there anything that's `AsRef<[u8]>` that isn't either `Deref<Target=[u8]>` or `Deref<Target=str>`?  (I couldn't see anything in std.)

**Comment by SimonSapin** [comment]
Regarding having a separate `write_string` function, I think that reducing the “API surface” is preferable when it doesn’t reduce functionality.

Regarding `read_utf8` vs `read_string`, it’s a trade-off between describing what the function does v.s. std-internal consistency with the existing `Read::read_to_string` method. I happen to put less value on internal consistency, but that’s really an opinion. The libs team’s consensus seems to be going the other way, so I’ll rename to `read_string`.

**Comment by alexcrichton** [comment]
@sfackler to confirm, has your concern with `write` been resolved?

**Comment by sfackler** [comment]
Yep, I think it's okay as-is.

**Comment by dtolnay** [comment]
@rfcbot resolved write

**Review by dtolnay — APPROVED** [review]
I see all the check boxes so let's get this merged. @SimonSapin please file a tracking issue and update the number in the unstable attributes. r=me

**Comment by SimonSapin** [comment]
Done: https://github.com/rust-lang/rust/issues/46588

**Comment by leonardo-m** [comment]
What's missing is an lazy iterator on text lines?

**Comment by Ryman** [comment]
I'm really happy about these additions but I'm just wondering if there is was an RFC which could be added as a reference to this issue?

**Comment by SimonSapin** [comment]
@Ryman This PR is closed (merged), further discussion should go into the tracking issue: https://github.com/rust-lang/rust/issues/46588

There was not RFC for this, but there will be a FCP before it is stabilized.

**Comment by Ryman** [comment]
@SimonSapin Thanks, I thought I was just really bad at searching :) I'll subscribe on there.

**Pull request merged**
